easy_profiler_converter/reader.cpp

```diff
@@ -0,0 +1,391 @@
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <iterator>
+
+///this
+#include "reader.h"
+
+#include "hashed_cstr.h"
+
+////from easy_profiler_core/reader.cpp/////
+
+typedef uint64_t processid_t;
+
+extern const uint32_t PROFILER_SIGNATURE = ('E' << 24) | ('a' << 16) | ('s' << 8) | 'y';
+
+# define EASY_VERSION_INT(v_major, v_minor, v_patch) ((static_cast<uint32_t>(v_major) << 24) | (static_cast<uint32_t>(v_minor) << 16) | static_cast<uint32_t>(v_patch))
+const uint32_t MIN_COMPATIBLE_VERSION = EASY_VERSION_INT(0, 1, 0); ///< minimal compatible version (.prof file format was not changed seriously since this version)
+const uint32_t EASY_V_100 = EASY_VERSION_INT(1, 0, 0); ///< in v1.0.0 some additional data were added into .prof file
+const uint32_t EASY_V_130 = EASY_VERSION_INT(1, 3, 0); ///< in v1.3.0 changed sizeof(thread_id_t) uint32_t -> uint64_t
+# undef EASY_VERSION_INT
+
+const uint64_t TIME_FACTOR = 1000000000ULL;
+
+// TODO: use 128 bit integer operations for better accuracy
+#define EASY_USE_FLOATING_POINT_CONVERSION
+
+#ifdef EASY_USE_FLOATING_POINT_CONVERSION
+
+// Suppress warnings about double to uint64 conversion
+# ifdef _MSC_VER
+#  pragma warning(disable:4244)
+# elif defined(__GNUC__)
+#  pragma GCC diagnostic push
+#  pragma GCC diagnostic ignored "-Wconversion"
+#  pragma GCC diagnostic ignored "-Wsign-conversion"
+# elif defined(__clang__)
+#  pragma clang diagnostic push
+#  pragma clang diagnostic ignored "-Wconversion"
+#  pragma clang diagnostic ignored "-Wsign-conversion"
+# endif
+
+# define EASY_CONVERT_TO_NANO(t, freq, factor) t *= factor
+
+#else
+
+# define EASY_CONVERT_TO_NANO(t, freq, factor) t *= TIME_FACTOR; t /= freq
+
+#endif
+
+inline bool isCompatibleVersion(uint32_t _version)
+{
+    return _version >= MIN_COMPATIBLE_VERSION;
+}
+
+#ifdef EASY_PROFILER_HASHED_CSTR_DEFINED
+
+typedef ::std::unordered_map<::profiler::block_id_t, ::profiler::BlockStatistics*, ::profiler::passthrough_hash<::profiler::block_id_t> > StatsMap;
+
+/** \note It is absolutely safe to use hashed_cstr (which simply stores pointer) because std::unordered_map,
+which uses it as a key, exists only inside fillTreesFromFile function. */
+typedef ::std::unordered_map<::profiler::hashed_cstr, ::profiler::block_id_t> IdMap;
+
+typedef ::std::unordered_map<::profiler::hashed_cstr, ::profiler::BlockStatistics*> CsStatsMap;
+
+#else
+
+// TODO: Create optimized version of profiler::hashed_cstr for Linux too.
+typedef ::std::unordered_map<::profiler::block_id_t, ::profiler::BlockStatistics*, ::profiler::passthrough_hash<::profiler::block_id_t> > StatsMap;
+typedef ::std::unordered_map<::profiler::hashed_stdstring, ::profiler::block_id_t> IdMap;
+typedef ::std::unordered_map<::profiler::hashed_stdstring, ::profiler::BlockStatistics*> CsStatsMap;
+
+#endif
+
+/// end from easy_profiler_core/reader.cpp/////
+
+using namespace profiler::reader;
+
+::profiler::block_index_t FileReader::readFile(const ::std::string &filename)
```

Comment:
Let's split this function on several functions

Reply:
ok. good idea, shall try.